File: packages/kotlin/src/test/kotlin/com/ecollect/sdk/ErrorMappingTest.kt
package com.ecollect.sdk

import com.ecollect.sdk.exceptions.*
import org.junit.Assert.*
import org.junit.Test

class ErrorMappingTest {

    @Test
    fun `SessionExpiredException is an EcollectException`() {
        val ex = SessionExpiredException("Session expired")
        assertTrue(ex is EcollectException)
        assertEquals("Session expired", ex.message)
    }

    @Test
    fun `InvalidConfigException is an EcollectException`() {
        val ex = InvalidConfigException("Invalid entity code")
        assertTrue(ex is EcollectException)
    }

    @Test
    fun `ValidationException is an EcollectException`() {
        val ex = ValidationException("Amount must be positive")
        assertTrue(ex is EcollectException)
    }

    @Test
    fun `InvalidCardException is an EcollectException`() {
        val ex = InvalidCardException("Card failed Luhn check")
        assertTrue(ex is EcollectException)
    }

    @Test
    fun `NetworkRetryableException is an EcollectException`() {
        val cause = RuntimeException("connection reset")
        val ex = NetworkRetryableException("Network error", cause)
        assertTrue(ex is EcollectException)
        assertEquals(cause, ex.cause)
    }

    @Test
    fun `TokenNotFoundException is an EcollectException`() {
        val ex = TokenNotFoundException("Token not found")
        assertTrue(ex is EcollectException)
    }

    @Test
    fun `DuplicateTransactionException is an EcollectException`() {
        val ex = DuplicateTransactionException("MerchantTransactionId already used")
        assertTrue(ex is EcollectException)
    }

    @Test
    fun `AuthenticationException is an EcollectException`() {
        val ex = AuthenticationException("Access denied")
        assertTrue(ex is EcollectException)
    }

    @Test
    fun `WebhookValidationException is an EcollectException`() {
        val ex = WebhookValidationException("SessionToken mismatch")
        assertTrue(ex is EcollectException)
    }

    @Test
    fun `PollingTimeoutException is an EcollectException`() {
        val ex = PollingTimeoutException("Polling timed out after 600s")
        assertTrue(ex is EcollectException)
    }

    @Test
    fun `CustomerNotFoundException is an EcollectException`() {
        val ex = CustomerNotFoundException("Customer not found")
        assertTrue(ex is EcollectException)
    }

    // Verify the return code -> exception mapping table
    @Test
    fun `return code mapping covers all critical error codes`() {
        // This test documents expected mappings for code review purposes
        val mappings = mapOf(
            "FAIL_APIEXPIREDSESSION" to "SessionExpiredException",
            "FAIL_INVALIDENTITYCODE" to "InvalidConfigException",
            "FAIL_INVALIDSERVICECODE" to "InvalidConfigException",
            "FAIL_INVALIDCREDITCARD" to "InvalidCardException",
            "FAIL_INVALIDEXPIRATIONDATE" to "InvalidCardException",
            "FAIL_TOKENNOTFOUND" to "TokenNotFoundException",
            "FAIL_TOKENEXPIRED" to "TokenNotFoundException",
            "FAIL_MERCHANTRANSID" to "DuplicateTransactionException",
            "FAIL_ACCESSDENIED" to "AuthenticationException",
            "FAIL_SESSIONNOTFOUND" to "WebhookValidationException",
            "FAIL_TICKETIDNOTMATCH" to "WebhookValidationException",
            "FAIL_SYSTEM" to "NetworkRetryableException"
        )

        assertEquals(12, mappings.size)
        assertTrue(mappings.containsKey("FAIL_APIEXPIREDSESSION"))
        assertTrue(mappings.containsKey("FAIL_SYSTEM"))
    }

    @Test
    fun `exponential backoff delays are correct`() {
        // delay = 2000 * 2^attempt
        val delays = (1..3).map { attempt -> 2000L * Math.pow(2.0, attempt.toDouble()).toLong() }
        assertEquals(4000L, delays[0])
        assertEquals(8000L, delays[1])
        assertEquals(16000L, delays[2])
    }
}

File: packages/php/tests/ErrorMappingTest.php
<?php

declare(strict_types=1);

namespace Ecollect\Tests;

use Ecollect\Exceptions\AuthenticationException;
use Ecollect\Exceptions\CustomerException;
use Ecollect\Exceptions\DuplicateTransactionException;
use Ecollect\Exceptions\EcollectException;
use Ecollect\Exceptions\InvalidCardException;
use Ecollect\Exceptions\InvalidConfigException;
use Ecollect\Exceptions\NetworkRetryableException;
use Ecollect\Exceptions\SessionExpiredException;
use Ecollect\Exceptions\TokenNotFoundException;
use Ecollect\Exceptions\ValidationException;
use Ecollect\Exceptions\WebhookValidationException;
use Ecollect\Utils\ErrorMapper;
use PHPUnit\Framework\TestCase;

class ErrorMappingTest extends TestCase
{
    /** @dataProvider errorMappingProvider */
    public function testErrorMapping(string $returnCode, string $expectedClass): void
    {
        $exception = ErrorMapper::map($returnCode);
        $this->assertInstanceOf($expectedClass, $exception);
        $this->assertSame($returnCode, $exception->getReturnCode());
    }

    public function errorMappingProvider(): array
    {
        return [
            ['FAIL_APIEXPIREDSESSION',    SessionExpiredException::class],
            ['FAIL_INVALIDENTITYCODE',    InvalidConfigException::class],
            ['FAIL_INVALIDSERVICECODE',   InvalidConfigException::class],
            ['FAIL_INVALIDCREDITCARD',    InvalidCardException::class],
            ['FAIL_INVALIDEXPIRATIONDATE', InvalidCardException::class],
            ['FAIL_TOKENNOTFOUND',        TokenNotFoundException::class],
            ['FAIL_TOKENEXPIRED',         TokenNotFoundException::class],
            ['FAIL_MERCHANTRANSID',       DuplicateTransactionException::class],
            ['FAIL_ACCESSDENIED',         AuthenticationException::class],
            ['FAIL_SESSIONNOTFOUND',      WebhookValidationException::class],
            ['FAIL_TICKETIDNOTMATCH',     WebhookValidationException::class],
            ['FAIL_SYSTEM',              NetworkRetryableException::class],
            ['FAIL_INVALIDTRANSVALUE',    ValidationException::class],
            ['FAIL_INVALIDCURRENCY',      ValidationException::class],
            ['FAIL_MAILFORMAT',           ValidationException::class],
            ['FAIL_CARDHOLDERID',         ValidationException::class],
            ['FAIL_CARDHOLDERIDTYPE',     ValidationException::class],
            ['FAIL_MOBILECOUNTRYCODE',    ValidationException::class],
            ['FAIL_MOBILENUMBER',         ValidationException::class],
            ['FAIL_CUSTOMERNOTFOUND',     CustomerException::class],
            ['FAIL_USERMISMATCH',         CustomerException::class],
        ];
    }

    public function testUnknownCodeReturnsBaseException(): void
    {
        $exception = ErrorMapper::map('FAIL_UNKNOWN_CODE');
        $this->assertInstanceOf(EcollectException::class, $exception);
        $this->assertStringContainsString('FAIL_UNKNOWN_CODE', $exception->getMessage());
    }

    public function testSystemErrorIsNetworkRetryable(): void
    {
        $exception = ErrorMapper::map('FAIL_SYSTEM', 'Custom error message');
        $this->assertInstanceOf(NetworkRetryableException::class, $exception);
    }

    public function testSessionExpiredHasCorrectCode(): void
    {
        $ex = ErrorMapper::map('FAIL_APIEXPIREDSESSION');
        $this->assertSame('SESSION_EXPIRED', $ex->getCodeStr());
    }
}
